R/theme.R

```diff
@@ -3,29 +3,111 @@ build_bslib <- function(pkg = ".") {
   bs_theme <- bs_theme(pkg)
 
   deps <- bslib::bs_theme_dependencies(bs_theme)
-  deps <- lapply(deps, htmltools::copyDependencyToDir, file.path(pkg$dst_path, "deps"))
+  deps <- lapply(deps, htmltools::copyDependencyToDir, path_deps(pkg))
   deps <- lapply(deps, htmltools::makeDependencyRelative, pkg$dst_path)
 
   head <- htmltools::renderDependencies(deps, srcType = "file")
-  write_lines(head, data_deps_path(pkg))
+
+  # include additional external assets
+  head <- paste(
+    head,
+    paste0(assemble_ext_assets(pkg), collapse = "\n"),
+    sep = "\n"
+  )
+
+  write_lines(head, path_data_deps(pkg))
+}
+
+assemble_ext_assets <- function(pkg) {
+  path_assets_yaml <- path_pkgdown(paste0("BS", pkg$bs_version), "assets_external.yaml")
+  deps_ext <- yaml::read_yaml(path_assets_yaml)
+
+  purrr::map_chr(deps_ext, ~ {
+    # download external resource
+    path <- path_deps(pkg, basename(.x$url))
+    download.file(.x$url, path, quiet = TRUE, mode = "wb")
+
+    # check file integrity
+    sha_size <- as.integer(regmatches(
+      .x$integrity,
+      regexpr("(?<=^sha)\\d{3}", .x$integrity, perl = TRUE)
+    ))
+    if (!(sha_size %in% c(256L, 384L, 512L))) {
+      cli::cli_abort(paste0(
+        "Invalid {.field integrity} value set in {.file ",
+        "{path_assets_yaml}}: {.val {(.x$integrity)}} Allowed are only ",
+        "SHA-256, SHA-384 and SHA-512."
+      ))
+    }
+    con <- file(path, encoding = "UTF-8")
+    hash <- openssl::base64_encode(openssl::sha2(con, sha_size))
+    hash_target <- regmatches(
+      .x$integrity,
+      regexpr("(?<=^sha\\d{3}-).+", .x$integrity, perl = TRUE)
+    )
+
+    if (hash != hash_target) {
+      cli::cli_abort(paste0(
+        "Hash of downloaded {(.x$type)} asset doesn't match {.field ",
+        "integrity} value of {.val {(.x$integrity)}}. Asset URL is: {.url ",
+        "{(.x$url)}}"
+      ))
+    }
+
+    # download subresources (webfonts etc.) if necessary
+    if (isTRUE(.x$has_subresources)) {
+      file_content <- read_file(path)
+      pos <- gregexpr("(?<=\\burl\\((?!(data|https?):))[^)?#]*", file_content, perl = TRUE)
+      urls <- unique(unlist(regmatches(file_content, pos)))
+      subdirs <- unique(fs::path_dir(urls))
+      fs::dir_create(
+        fs::path_norm(fs::path(path_deps(pkg), subdirs)),
+        recurse = TRUE
+      )
+      url_parsed <- xml2::url_parse(.x$url)
+      url_excl_scheme <- fs::path(url_parsed$server, url_parsed$path)
+      remote_urls <- paste0(
+        url_parsed$scheme, "://",
+        fs::path_norm(fs::path(fs::path_dir(url_excl_scheme), urls))
+      )
+      purrr::walk2(
+        remote_urls,
+        urls,
+        ~ download.file(.x, fs::path_norm(path_deps(pkg, .y)), quiet = TRUE, mode = "wb")
```

Comment:
It seems like these get downloaded on demand every time you build the site? Would it be better to bundle them in pkgdown itself, and then just provide some function to update pkgdown when needed?

Reply:
Yes, the subresources get re-downloaded on every build (same as the main asset resources). Bundling them would make pkgdown exceed CRAN's package size limits by far [as you noticed some while ago](https://github.com/r-lib/pkgdown/pull/1541#issuecomment-792732020). Instead, I could implement a simple caching approach, i.e. only re-download a file if it doesn't already exist in destination folder, taking the file integrity hash into account for the main resources. The subresources are currently not explicltly recorded in `[inst/BS5/assets_external.yaml](https://github.com/r-lib/pkgdown/pull/2249/files#diff-07268511efc6e5849184d97100ceb744bc64a328ab6b3ce97e1fb7e5bfd7dd94)` and hence we don't have integrity hashes for them. So for subresources I would simply check if the file already exists and skip re-downloading if it does. One could always delete the pkgdown output to trigger a full re-download of all assets if needed.

That would actually be easy to implement. What do you think?

Reply:
What are you thinking to use for the file paths? Just the name of the resource or the name and version?

Reply:
Maybe it would be better to not download from the CDN and just install directly from their github etc. That way we shouldn't have to worry about them putting subresoures in such weird locations?

Reply:
That could be better, I guess, since then the project could set up [dependabot](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates) or similar to automatically receive update PRs for the included assets. But it would need a lot more work (none of which is found in this PR).

Regarding the "weird locations" I'm not so sure. Might be that path hierarchy of interdependent files in assets like Font Awesome icons is kinda hardcoded, i.e. not easily customizable. I dunno.

Reply:
Also beware that we don't want to include the actual assets themselves but only some kind of reference to them so they're downloaded on demand during pkgdown build. I.e. links to the GitHub artifacts of the built assets or something. Seems not trivial to me.

Reply:
Maybe we can just find a different approach for font awesome then? I don't think we need to create a whole system for dealing with subresources if we can figure out how to handle font awesome.

Also feel free to just leave this and I'll pick it up again when I'm next working on pkgdown.

Reply:
Font Awesome assets could also be copied from the R package [fontawesome](https://rstudio.github.io/fontawesome/), I guess? It ships all the required CSS files + fonts, [it appears](https://github.com/rstudio/fontawesome/tree/main/inst/fontawesome)...

Only downside: one more package dependency.

Reply:
We're depending on fontawsome either way; if it's already bundled in a package we might as well use it.

Reply:
So, to sum: Do we agree the following changes need to be done to this PR to land in pkgdown?

1. Re-download assets only if they don't already exist in destination folder (taking their hash into account) [as outlined above](https://github.com/r-lib/pkgdown/pull/2249#discussion_r1189928290).

2. Remove subresources logic and handle Font Awesome icons separately, i.e. copy them from R pkg `fontawesome`.

Or do you want to re-implement asset inclusion differently / from scratch?